contrib/babelfishpg_common/src/numeric.c

```diff
@@ -1029,3 +1064,253 @@ tsql_numeric_get_typmod(Numeric num)
 
 	return (((precision & 0xFFFF) << 16 ) | (scale & 0xFFFF)) + VARHDRSZ;
 }
+
+Datum
+bigint_poly_sum(PG_FUNCTION_ARGS)
+{
+	
+	PolyNumAggState *state;
+	
+	state = PG_ARGISNULL(0) ? NULL : (PolyNumAggState *) PG_GETARG_POINTER(0);
+	/* If there were no non-null inputs, return NULL */
+	if (state == NULL || state->N == 0)
+		PG_RETURN_NULL();
+
+	#ifdef HAVE_INT128
+	int128 			arg;
+	arg = state->sumX;
+
+	if (unlikely(arg < PG_INT64_MIN) || unlikely(arg > PG_INT64_MAX))
+		ereport(ERROR,
+				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				 errmsg("Arithmetic overflow error converting expression to data type bigint.")));
+
+	PG_RETURN_INT64((int64) arg);
+	#else
+
+	Datum temp;
+	bool flag;
+	int64 result;
+	NumericVar nvar;
+
+	temp = DirectFunctionCall1(numeric_sum, PG_GETARG_DATUM(0));
+	init_var(&nvar);
+	set_var_from_num( DatumGetNumeric(temp), &nvar);
+	flag = numericvar_to_int64( &nvar, &result);
+	free_var(&nvar);
+
+	if (!flag)
+		ereport(ERROR,
+				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				 errmsg("Arithmetic overflow error converting expression to data type bigint.")));
+	else
+	PG_RETURN_INT64(result);
+	#endif
+}
+
+Datum
+int4int2_sum(PG_FUNCTION_ARGS)
+{
+
+	int64 result;
+ 	if (PG_ARGISNULL(0))
+	{
+		PG_RETURN_NULL();
+	}
+	else
+	{
+		result  = PG_GETARG_INT64(0);
+
+		if (unlikely(result < PG_INT32_MIN) || unlikely(result > PG_INT32_MAX))
+		ereport(ERROR,
+				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				 errmsg("Arithmetic overflow error converting expression to data type int.")));
+
+		PG_RETURN_INT32((int32) result);
+	}
+}
+
+static bool
```

Comment:
Add comment here that this is taken from numeric.c and explain the return value also,
```
/*
 * Convert numeric to int8, rounding if needed.
 *
 * If overflow, return false (no error is raised).  Return true if okay.
 */
```

Reply:
Added required comment to all the functions referenced/used